internal/c/parts/video/font/font.cpp

```diff
@@ -1,339 +1,1417 @@
-                            int32_t *out_x_pre_increment, int32_t *out_x_post_increment) {
-    static uint32_t *utf32_codepoint;
-    static int32_t retval;
-    if (codepoints >= 1) {
-        utf32_codepoint = (uint32_t *)malloc(codepoints * 4);
-        static int32_t x;
-        for (x = 0; x < codepoints; x++) {
-            utf32_codepoint[x] = codepage437_to_unicode16[codepoint[x]];
-        }
-    }
-    retval = FontRenderTextUTF32(i, utf32_codepoint, codepoints, options, out_data, out_x, out_y, out_x_pre_increment, out_x_post_increment);
-    if (codepoints > 0)
-        free(utf32_codepoint);
-    return retval;
+/// @brief Returns the vertical line spacing in pixels (font height + extra pixels if any)
+/// @param qb64_fh A QB64 font handle (this can be a builtin font as well)
+/// @param passed Optional arguments flag
+/// @return The vertical spacing in pixels
+int32_t func__ULineSpacing(int32_t qb64_fh, int32_t passed) {
+    if (new_error)
+        return 0;
+
+    if (passed) {
+        // Check if a valid font handle was passed
+        if (qb64_fh != 8 && qb64_fh != 14 && qb64_fh != 16 && !(qb64_fh >= 32 && qb64_fh <= lastfont && font[qb64_fh])) {
+            error(258);
+            return 0;
+        }
+    } else {
+        qb64_fh = write_page->font; // else get the current write page font handle
+    }
+
+    // For buint-in fonts return the handle value (which is = font height)
+    if (qb64_fh < 32)
+        return qb64_fh;
+
+    FONT_DEBUG_CHECK(IS_FONT_HANDLE_VALID(font[qb64_fh]));
+
+    auto fnt = fontManager.fonts[font[qb64_fh]];
+    auto face = fnt->face;
+
+    return lroundf(((float)(face->height) / (float)face->units_per_EM * (float)fnt->defaultHeight) + 2.0f);
 }
 
-int32_t FontWidth(int32_t i) {
-    if (fonts[i].monospace)
-        return fonts[i].monospace_width;
-    return 0;
+/// @brief This renders text on an active destination (graphics mode only) using the currently selected color
+/// @param start_x The starting x position
+/// @param start_y The starting y position
+/// @param text The text that needs to be rendered
+/// @param max_width The maximum width of the text (rendering will be clipped beyond width)
+/// @param utf_encoding The UTF encoding of the text (0 = ASCII, 8 = UTF-8, 16 - UTF-16, 32 = UTF-32)
+/// @param qb64_fh A QB64 font handle (this can be a builtin font as well)
+/// @param passed Optional arguments flag
+void sub__UPrintString(int32_t start_x, int32_t start_y, const qbs *text, int32_t max_width, int32_t utf_encoding, int32_t qb64_fh, int32_t passed) {
+    if (new_error || !text->len)
+        return;
+
+    // Check if we are in text mode and generate an error if we are
+    if (write_page->text) {
+        error(5);
+        return;
+    }
+
+    // Check max width
+    if (passed & 1) {
+        if (max_width < 1)
+            return;
+    } else {
+        max_width = 0;
+    }
+
+    // Check UTF argument
+    if (passed & 2) {
```

Comment:
It seems like a lot of this logic is common among a few of the commands and could be moved into a separate function.

Reply:
Moved these into macros.